### ISSUE TITLE
Accept FormData as input

### ISSUE THREAD

**Issue by rclmenezes** [TheEdoRan/next-safe-action#2]
[Most of Next's docs](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions) are passing server actions to the `action` parameter of `form`:

```typescript
import { cookies } from 'next/headers';
 
export default function AddToCart({ productId }) {
  async function addItem(data) {
    'use server';
 
    const cartId = cookies().get('cartId')?.value;
    await saveToDb({ cartId, data });
  }
 
  return (
    <form action={addItem}>
      <button type="submit">Add to Cart</button>
    </form>
  );
}
```

However, `next-safe-action`'s `ClientAction` doesn't take FormData as an input. This means that we have to do a lot of casting to make actions work:

```typescript
    <form
      action={async (formData) => {
        "use server";

        await loginUser({
          username: formData.get("username") as string,
          password: formData.get("password") as string,
        });
      }}
    >
```

Maybe we could make it accept `FormData` as an input and let Zod do the work?

```typescript
    <form
      action={async (formData) => {
        "use server";

        await loginUser(formData);
      }}
    >
```

**Comment by TheEdoRan** [comment]
Thanks for opening this issue.
As you can see [here](https://github.com/TheEdoRan/next-safe-action-example/blob/main/src/app/login-form.tsx#L21), you can use `Object.fromEntries` to convert `FormData` into object. Then you can pass it as action input. You can also cast the object to `any` or `z.infer<typeof LoginValidator>` (assuming the input validator for your Server Action is called `LoginValidator`) if you don't want to specify key names.
I think implementing this in the library while keeping typesafe input could be a little difficult. I absolutely wanna keep the typesafe input, because you can use Server Actions outside of forms.

Let me know what you think about this solution.

**Comment by rclmenezes** [comment]
> you can use Object.fromEntries to convert FormData into object

Fair point!

> I absolutely wanna keep the typesafe input, because you can use Server Actions outside of forms.

Personally, I'm not sure I'd use a safeAction in the server that way. After all, if the inputs are already validated then we don't need Zod to run.

One alternative would be something like:

```typescript
// Use this in the server because you don't need Zod
export const loginUser = async ({ username, password }) => {
    ...
};

// Use this as a server action because we need to validate input
export const loginUserAction = action({ input }, loginUser);
```

It's a little clunky, but this way Zod doesn't run on already validated input.

But your call :)

**Comment by TheEdoRan** [comment]
Hmm, I'm not sure I understand the problem. Zod runs once per action call on the server, and it validates the incoming input from client. This validation is necessary because the action is a simple POST request, and someone can even make it outside of Next.

What I mean by "typesafe input" is: on the client, when you call the action, as a param you need to pass an input object. That input object has the same type as the inferred (`z.infer`) input validator. This is nice because this way we know what to pass to the action from client, but **this isn't validation**, it's just TypeScript types, i.e. you can pass wrong data (for example by casting input to `any` ), and that way TS doesn't complain. This is why Zod is needed: TS is just a build time "guard", it doesn't validate input at runtime.

So, TL;DR: it's good DX because you know what to pass as input on client thanks to TypeScript, but then Zod needs to do the validation work, 'cause we don't trust client requests.

Sorry if I misunderstood your comment. If that's the case, please let me know!

**Comment by rclmenezes** [comment]
I think _I_ may have misunderstood your comment.

When you said "I absolutely wanna keep the typesafe input, because you can use Server Actions outside of forms" I thought you were saying 'I want to be able to use server actions in non-action scenarios'. For example, a "createUser" server action function could theoretically be used in a script to import users.

That's the part I disagree with. In that scenario, you don't really need Zod. That's why I was pointing out that you could separate the action from the safe action in separate exports.

That said -- I now agree with you on the core issue of strongly typing server action inputs. By forcing people to define their inputs, we ensure that if the server action arguments change, users are reminded to change the form as well. It solves a different problem than Zod. If they want to opt out with an `any`, they can.

I think the pattern in the Nextjs docs of passing formData in blindly is going to be rare. Honestly, most people are probably going to use `react-hook-form` additionally any way.

Anyway, thanks for the conversation 🙏
